src/algebra/char_p/mixed_char_zero.lean

```diff
@@ -0,0 +1,452 @@
+      end,
+      rw eq_comm,
+      apply or_iff_not_imp_left.mp (nat.prime.eq_one_or_self_of_dvd hp r r_dvd_p),
+      exact char_p.char_ne_one (R ⧸ M) r }},
+  { intro g,
+    rcases g with ⟨I, ⟨hI_max, hI⟩⟩,
+    use I,
+    exact ⟨ideal.is_maximal.ne_top hI_max, hI⟩ }
+end
+
+end mixed_char_zero
+
+/-!
+### Equal characteristic zero
+
+A commutative ring `R` has "equal characteristic zero" if it satisfies one of the following
+equivalent properties:
+
+1) `R` is a `ℚ`-algebra.
+2) The quotient `R ⧸ I` has characteristic zero for any proper ideal `I ⊂ R`.
+3) `R` has characteristic zero and does not have mixed characteristic for any prime `p`.
+
+We show `(1) ↔ (2) ↔ (3)`, and most of the following is concerned with constructing
+an explicit algebra map `ℚ →+* R` (given by `x ↦ (x.num : R) /ₚ ↑x.pnat_denom`)
+for the direction `(1) ← (2)`.
+
+Note: Property `(2)` is denoted as `equal_char_zero` in the statement names below.
+-/
+section equal_char_zero
+
+/--
+`ℚ`-algebra implies equal characteristic.
+-/
+lemma Q_algebra_to_equal_char_zero [nontrivial R] [algebra ℚ R] :
+  ∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I) :=
+begin
+    haveI : char_zero R := @char_p.char_p_to_char_zero R _
+      (char_p_of_injective_algebra_map (algebra_map ℚ R).injective 0),
+    intros I hI,
+    exact
+    ⟨begin
+      intros a b h_ab,
+      apply @nat.cast_injective R,
+      simp_rw ←map_nat_cast (ideal.quotient.mk I) at h_ab,
+      simp_rw ←map_nat_cast (algebra_map ℚ R) at ⊢ h_ab,
+      apply congr_arg,
+      rw ←sub_eq_zero at ⊢ h_ab,
+      simp_rw ←map_sub at ⊢ h_ab,
+      rw ideal.quotient.eq_zero_iff_mem at h_ab,
+
+      -- `i(↑a - ↑b)` is a unit contained in `I`, which contradicts `I ≠ ⊤`.
+      by_contradiction h_ne_zero,
+      have hI' : I = ⊤ :=
+        I.eq_top_of_is_unit_mem h_ab (is_unit.map (algebra_map ℚ R) (is_unit.mk0 _ h_ne_zero)),
+      exact absurd hI' hI,
+    end⟩
+end
+
+section construction_of_Q_algebra
+
+/-- Internal: Not intended to be used outside this local construction. -/
+lemma equal_char_zero.pnat_coe_is_unit [h : fact (∀ (I : ideal R), I ≠ ⊤ → char_zero (R ⧸ I))]
+  (n : ℕ+) : is_unit (n : R) :=
+begin
+  rw is_unit_iff_exists_inv',
+
+  -- Assume `n` is not invertible.
+  by_contradiction h',
+  rw not_exists at h',
+
+  -- Then `(n)` is a proper ideal in `R`.
+  let I := ideal.span ({n} : set R),
+  have hI : (1 : R) ∉ I :=
+  begin
+    by_contradiction one_mem_I,
+    cases ideal.mem_span_singleton'.mp one_mem_I with x hx,
+    exact absurd hx (h' x),
+  end,
+
+  -- by assumption `char_zero R⧸(n)` so `n = 0`.
+  have h_char_zero : char_zero (R ⧸ I) := (h.elim) I ((ideal.ne_top_iff_one I).mpr hI),
+  have n_zero : (n : ℕ) = 0 :=
+  begin
+    apply h_char_zero.cast_injective,
+    rw ←map_nat_cast (ideal.quotient.mk I),
+    rw [nat.cast_zero, ideal.quotient.eq_zero_iff_mem, ideal.mem_span_singleton'],
+    use 1,
+    simp,
+  end,
+  exact absurd n_zero (ne_of_gt n.property),
+end
```

Comment:
I made a few of the proofs by contradiction into direct proofs and used some handy lemmas to golf this:
```suggestion
begin
  -- `n : R` is a unit iff `(n)` is not a proper ideal in `R`.
  rw ← ideal.span_singleton_eq_top,
  -- So by contrapositive, we should show the quotient does not have characteristic zero.
  apply not_imp_comm.mp (h.elim (ideal.span {n})),
  unfreezingI { intro h_char_zero },
  -- In particular, the image of `n` in the quotient should be nonzero.
  apply (h_char_zero.cast_injective).ne n.ne_zero,
  -- But `n` generates the ideal, so its image is clearly zero.
  rw [←map_nat_cast (ideal.quotient.mk _), nat.cast_zero, ideal.quotient.eq_zero_iff_mem],
  exact ideal.subset_span (set.mem_singleton _)
end
```